---
title: "'stringr' and 'data.table': Wrangling data big and small - not rocket science" 
author: "Kate Crosby"
font-import: http://fonts.googleapis.com/css?family=Lato
font-family: 'Lato'
date: "November 19, 2014"
output: ioslides_presentation
---

## Two COMMON problems

- Arrays of data are: <div class="red3">__BIG__</div>

- <div class="red3">__OTHER PEOPLE/MACHINES__ </div> formattted it or typed it (not always correctly)

## Two potential solutions in the form of R-packages

1. <div class="red3">'data.table'</div>
    + subset, modify, destroy, query, merge large datatables
    + understandable and easy syntax
2. <div class="red3">'stringr'</div> 
    + R not noted for its handling of strings
    + remove annoying spaces, underscores, etc
    + use regular expressions
    
THANKS HADLEY!

## Example 1 - read in a 'large' file with data.table 'fread' (won't take gzip nicely)

```{r, echo=TRUE}
library(data.table)
system.time(fulldata <- fread("NAMZeaGBSv2.3_allChrs.hmp.txt",
                  header=T, 
                  stringsAsFactors=F, sep ="\t"))
dim(fulldata)
```

## Contrasting example 1 - read in a 'large' file with read.table (slower)

```{r, echo=TRUE}
system.time(table.diff <- read.table("NAMZeaGBSv2.3_allChrs.hmp.txt",
                  header=T, comment.char="",stringsAsFactors=F))
dim(table.diff)

```

## Example 2 - I have too many columns/rows! Deletion all at once

- Very handy := **NULL**
```{r, echo = TRUE}
system.time(fulldata_ped <- fulldata[,c('alleles', 'chrom', 
                                        'pos', 'cm', 'assembly#', 
                                        'center','protLSID',
                                        'assayLSID', 'panelLSID', 
                                        'QCcode') := NULL])
dim(fulldata_ped)
```

## Contrasting example 2 - Exclusion subset a dataframe traditionally
- Two commmands, still fast, but still obnoxious specification
```{r, echo = TRUE}
drops <- c("alleles", "chrom", "pos", "cm", 
           "assembly#", 
           "center", "protLSID",
           "assayLSID", "panelLSID", 
           "QCcode")
system.time(new.df <- table.diff[,!(names(table.diff) %in% drops)])
```

## Another example - subset with select (just like base) = for INCLUSION

```{r, echo=FALSE}
fulldata <- fread("NAMZeaGBSv2.3_allChrs.hmp.txt",
                  header=T, stringsAsFactors=F, sep ="\t")
```

```{r, echo=TRUE}
fulldata_map <- subset(fulldata, select = c("chrom", "rs#", "pos"))
head(fulldata_map)
```

## Order data really quickly with setkey() - here by strand then chromosome
```{r, echo=FALSE}
fulldata <- fread("NAMZeaGBSv2.3_allChrs.hmp.txt",
                  header=T, stringsAsFactors=F, sep ="\t")
fulldata_map <- subset(fulldata, select = c("chrom", "rs#", "pos", "cm"))
```
```{r, echo=TRUE}
order.dt <- setkey(fulldata_map, "cm", "chrom")
head(order.dt)
```
## **.SD** = "S ubset of D ata.table"
Unreasonable requests: "*Give me the 2nd row (SNP) for/of each chromosome!*"
```{r, echo = FALSE}
fulldata_ped <- fulldata[,c('pos', 'alleles', 'cm', 'assembly#', 
                            'center','protLSID','assayLSID', 
                            'panelLSID', 'QCcode') := NULL]
```
```{r, echo = TRUE}
x <-fulldata_ped[,.SD[2], by = chrom] 
dim(x)
head(x)
```

## Phenotype data with stringr 
```{r, echo = TRUE}
phenos <- fread("phenos.txt", header=T, stringsAsFactors=F, sep ="\t")
x <- subset(phenos, select = c("PHENOTYPE_VALUE", "PHENOTYPE_NAME", 
                                "ACCESSION"))
head(x)

```

## Get rid of 'measure=' but not the value
```{r, echo = TRUE}
library(stringr)
pheno.new <- str_replace(phenos$PHENOTYPE_VALUE, "measure=", "")

head(pheno.new)
```
## Get rid of ';' but not the value
```{r, echo = TRUE}
pheno.new <- str_replace(pheno.new, ";", "")
head(pheno.new)

```

## More stringr: Looking for a weird restriction site **TGCGC**
```{r, echo = TRUE}
DNA <- c("GCTAAGCCCAATTTATGCGC", "GATAAGCCCTGCGCAATA")
# Return it to me in 3 pieces - third piece is empty
str_split_fixed(DNA, "TGCGC", 3)
# Cut in two instead
str_split_fixed(DNA, "TGCGC", 2)
```

## Query or get the mean phenotypes by trait or by anything else 
```{r, echo = T}
phenos <- fread("pheno.new.txt", header=T, stringsAsFactors=F, sep ="\t")
DT <- subset(phenos, select = c("phenoV", "PHENOTYPE_NAME", 
                                "ACCESSION", "STATE"))
head(DT)

```
## Query or get the mean phenotypes by trait or by anything else 
```{r, echo = T}
x <- DT[,mean(phenoV, na.rm=T), by = 'PHENOTYPE_NAME']
head(x)
```
## Another nice example is merging large data.tables by a common attribute
- Let's merge a pedigree file with a large GRIN dataset, ordered using setkey() by 
Accession number
- Ordering by setkey() makes this merging very, very, fast
```{r, echo =FALSE}
# import the zea GRIN dataframe - includes more than just Zea mays
grin.all <- read.csv("GRIN.csv", header = T, stringsAsFactors = F)

# import the pedigree dataframe of interest
amesWyears <- read.csv("ames_withyears.csv", header = T, stringsAsFactors = F)


# load libraries
library(stringr)
library(data.table)

# replace the space in GRIN 
# accession number so that it matches pedigree csv Accesion.N or whatever field

accenumb <- grin.all$accenumb

accenumbPI <- str_replace(accenumb, "I ", "I")
accenumbMBG<- str_replace(accenumbPI, "S ", "S")
accenumbAmes <- str_replace(accenumbMBG, "s ", "s")
accenumbNSL <- str_replace(accenumbAmes, "L ", "L")

# Rename that new vector so that it matches pedigree files - 
# there still a few ones with weird spaces, can change later
Accesion.N <- accenumbNSL

# New dataframe
grin.new <- data.frame(grin.all, Accesion.N, stringsAsFactors = F)

```
```{r, echo=TRUE}

zea.grin.T <- data.table(grin.new)
amesWyears.T <- data.table(amesWyears)
setkey(zea.grin.T, Accesion.N)
setkey(amesWyears.T, Accesion.N)

system.time(new.T <- zea.grin.T[amesWyears.T, roll =T])

```

## Look at the new table, and subset for simplicity

```{r, echo = T}
sdT <- subset(new.T, select=c('Inbred.line', 'Year', 
                              'released', 
                              'Accesion.N', 'accenumb', 'ancest'))
sdT[1:10,]
```

